### ISSUE TITLE
[Bug]userFunction.apply is not a function, when using a recursive loop

### ISSUE THREAD

**Issue by oosugi84** [appsmithorg/appsmith#6946]
## Description

[What happened]
I tried to chaining actions with a loop, the code is:
```
{{
  (function() {
      function loopArray(arr, idx) {
          preprocess(idx, arr[idx], function(index) {
            console.log('in callback')
              if(index < arr.length) {
                  loopArray(arr, index);
              }
          });
      }
      function preprocess(i, col, callback) {
          let idx = i;
          preProcessApi.run((response, params) => {
              console.log(params);
              params.idx += 1;
              storeValue("preprocessedNum", params.idx, false);
              callback(params.idx);
          }, (response, params) => { 
              params.idx += 1;
              storeValue("preprocessedNum", params.idx, false);
              callback(params.idx);
          }, { id: col.ID, idx } )
      }
      loopArray(Table1.tableData, 0);
  })()
}}
```
When it runs to the line `callback(params.idx);`
There is an exception raised. Error occurred while evaluating the trigger: TypeError: userFunction.apply is not a function

Expect:
Loop the tableData with actions which is one after another.

**Comment by oosugi84** [comment]
The code could be neat and tidy if async await is supported.

**Comment by Nikhil-Nandagopal** [comment]
Solved with #8988
